### PR TITLE
Error when creating qnode with backprop and finite-shots

### PR DESCRIPTION
**Pull request by albi3ro**
Fixes Issue #1481 

Right now, you can create a qnode that uses finite shots and backpropagation.  This is fine on qnode execution on the forward pass, but will always fail and raise some sort of unhelpful error message of the backward pass/ gradient computation.

For example:

```python
dev = qml.device('default.qubit.autograd', shots=10, wires=1)

@qml.qnode(dev, interface='autograd', diff_method='backprop')
def circ(x):
    qml.RX(x, wires=0)
    return qml.expval(qml.PauliZ(0))

## works fine
circ(0.1)

# incomprehensible error
qml.grad(circ)(0.1)
```

For context, that error is:
```
ValueError: setting an array element with a sequence.
```

With this change, during QNode creation, a `UserWarning` is raise.  I decided to use a warning instead of a Error, as users may still want to calculate finite shot forward passes on an interface device.  This way, they at least have an explanation for the strange errors they will end up getting.

### CODE REVIEW

**Comment by antalszava** [comment]
[ch8003]

**Comment by antalszava** [comment]
Hi @albi3ro what's the status on this?

> I decided to use a warning instead of a Error, as users may still want to calculate finite shot forward passes on an interface device.

Could they simply just evaluate the QNode without trying to differentiate it? If so, maybe we could refer to that in the error message.

File: pennylane/qnode.py [hunk 1]
```diff
@@ -342,6 +342,12 @@ def _validate_backprop_method(device, interface):
             qml.QuantumFunctionError: if the device does not support backpropagation, or the
             interface provided is not compatible with the device
         """
+        if device.shots is not None:
```
Comment (rmoyard) [inline pennylane/qnode.py:345]: Great 👍

File: tests/tape/test_qnode.py [hunk 2]
```diff
@@ -70,6 +70,15 @@ def test_validate_device_method(self, monkeypatch):
         assert interface == "interface"
         assert device is dev
 
+    @pytest.mark.parametrize("interface", ("autograd", "torch", "tensorflow", "jax"))
+    def test_validate_backprop_method_finite_shots(self, interface):
+        """Tests that an error is raised for backpropagation with finite shots."""
+
+        dev = qml.device("default.qubit", wires=1, shots=3)
+
+        with pytest.raises(qml.QuantumFunctionError, match="Devices with finite shots"):
```
Comment (rmoyard) [inline tests/tape/test_qnode.py:79]: Nice test 💯 do you think codecov is wrong?

File: tests/devices/test_default_qubit_tf.py [hunk 3]
```diff
@@ -1329,23 +1329,18 @@ def test_sample_observables_non_differentiable(self):
         shots = 100
         dev = qml.device("default.qubit.tf", wires=2, shots=shots)
 
-        @qml.qnode(dev, diff_method="backprop", interface="tf")
-        def circuit(a):
-            qml.RX(a, wires=0)
-            return qml.sample(qml.PauliZ(0))
-
-        a = tf.Variable(0.54)
+        with pytest.raises(qml.QuantumFunctionError, match="Devices with finite shots"):
```
Comment (rmoyard) [inline tests/devices/test_default_qubit_tf.py:1332]: Should we keep this test? Maybe changing the description to better fit the error message?

**Review by rmoyard — COMMENTED** [review]
Great, I've left some comments. 💯  I think we should add a new test, because a deleted test that was previously covering some jax functionalities was deleted.

File: tests/devices/test_default_qubit_tf.py (hunk 3 above)
Comment (albi3ro) [reply tests/devices/test_default_qubit_tf.py:1332]: Since we test elsewhere that this error is raised, I'm deleting this test.

**Comment by albi3ro** [comment]
@rmoyard  what do we need to add?

**Comment by rmoyard** [comment]
@albi3ro I think one one the test using jax was deleted. `default_qubit_jax.py` line 269 more precisely we should add something covering this line.

File: tests/devices/test_default_qubit_jax.py [hunk 4]
```diff
@@ -152,14 +152,15 @@ def circuit():
         expected = jnp.array([amplitude, 0, jnp.conj(amplitude), 0])
         assert jnp.allclose(state, expected, atol=tol, rtol=0)
 
+    @pytest.mark.skip(reason="sampling doesnt work in backprop, fails with parameter-shift")
```
Comment (albi3ro) [inline tests/devices/test_default_qubit_jax.py:155]:
```suggestion
```

**Review by rmoyard — APPROVED** [review]
Great, it seems good to me 💯

**Pull request merged**